File: tools/ws_test_server.js
#!/usr/bin/env node
/**
 * Minimal WebSocket test server for mod_audio_stream (Node.js).
 *
 * - Logs incoming text/binary messages.
 * - Echoes text back in a JSON envelope.
 * - Optionally sends a streamAudio JSON with a provided raw PCM file (16-bit LE).
 *
 * Usage:
 *   node tools/ws_test_server.js --port 8765
 *   node tools/ws_test_server.js --port 8765 --send-raw sample.raw --rate 8000 --channels 1
 *
 * Requires: npm install ws
 */

const fs = require('fs');
const path = require('path');
const WebSocket = require('ws');

function parseArgs() {
  const args = {
    host: '0.0.0.0',
    port: 8765,
    sendRaw: null,
    rate: 8000,
    channels: 1,
    frameMs: 20,
    help: false
  };

  const argv = process.argv.slice(2);
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--help' || a === '-h') {
      args.help = true;
    } else if (a === '--port' && argv[i + 1]) {
      args.port = parseInt(argv[++i], 10);
    } else if (a === '--host' && argv[i + 1]) {
      args.host = argv[++i];
    } else if (a === '--send-raw' && argv[i + 1]) {
      args.sendRaw = argv[++i];
    } else if (a === '--rate' && argv[i + 1]) {
      args.rate = parseInt(argv[++i], 10);
    } else if (a === '--channels' && argv[i + 1]) {
      args.channels = parseInt(argv[++i], 10);
    } else if (a === '--frame-ms' && argv[i + 1]) {
      args.frameMs = parseInt(argv[++i], 10);
    }
  }
  return args;
}

function printUsage() {
  console.log(`Usage:
  node tools/ws_test_server.js [--host 0.0.0.0] [--port 8765] [--send-raw file.raw] [--rate 8000] [--channels 1] [--frame-ms 20]

Examples:
  node tools/ws_test_server.js --port 8765
  node tools/ws_test_server.js --port 8765 --send-raw sample.raw --rate 16000 --channels 1 --frame-ms 20`);
}

async function main() {
  const opts = parseArgs();
  if (opts.help) {
    printUsage();
    process.exit(0);
  }

  let rawPayload = null;
  if (opts.sendRaw) {
    try {
      const buf = fs.readFileSync(path.resolve(opts.sendRaw));
      rawPayload = buf.toString('base64');
      console.log(`[init] loaded raw file ${opts.sendRaw} (${buf.length} bytes)`);
    } catch (err) {
      console.error(`[init] failed to read raw file: ${err.message}`);
    }
  }

  const wss = new WebSocket.Server({ host: opts.host, port: opts.port }, () => {
    console.log(`[*] listening on ws://${opts.host}:${opts.port}`);
  });

  wss.on('connection', (ws, req) => {
    const peer = req.socket.remoteAddress + ':' + req.socket.remotePort;
    console.log(`[+] connection from ${peer}`);

    let seq = 0;
    let upstreamRate = opts.rate || 8000;
    let upstreamChannels = opts.channels || 1;
    const frameMs = opts.frameMs || 20;

    if (rawPayload) {
      const payload = {
        type: 'streamAudio',
        data: {
          audioDataType: 'raw',
          sampleRate: upstreamRate,
          channels: upstreamChannels,
          audioData: rawPayload
        }
      };
      ws.send(JSON.stringify(payload));
      console.log(`[>] sent streamAudio: bytes=${Buffer.from(rawPayload, 'base64').length} rate=${upstreamRate} ch=${upstreamChannels}`);
    }

    ws.on('message', (data, isBinary) => {
      if (isBinary) {
        // если знаем длительность кадра, попробуем оценить частоту дискретизации
        const samplesPerFrame = data.length / 2 / upstreamChannels;
        const guessedRate = Math.round(samplesPerFrame * (1000 / frameMs));
        if (guessedRate > 0 && guessedRate !== upstreamRate) {
          upstreamRate = guessedRate;
        }
        console.log(`[<] binary: ${data.length} bytes (rate≈${upstreamRate} Hz, ch=${upstreamChannels})`);
        // Эхо назад как streamAudio raw с теми же rate/ch
        const payload = {
          type: 'streamAudio',
          data: {
            audioDataType: 'raw',
            sampleRate: upstreamRate,
            channels: upstreamChannels,
            audioData: data.toString('base64'),
            seq: seq++
          }
        };
        ws.send(JSON.stringify(payload));
        return;
      }
      const text = data.toString();
      console.log(`[<] text: ${text.substring(0, 200)}`);
      try {
        const msg = JSON.parse(text);
        if (msg.type === 'rawAudio' && msg.data) {
          if (msg.data.sampleRate) upstreamRate = msg.data.sampleRate;
          if (msg.data.channels) upstreamChannels = msg.data.channels;
          console.log(`[meta] rawAudio from FS: rate=${upstreamRate} ch=${upstreamChannels}`);
        }
      } catch (e) {
        const reply = { echo: text };
        ws.send(JSON.stringify(reply));
        console.log('[>] echo reply (non-JSON)');
      }
    });

    ws.on('close', (code, reason) => {
      console.log(`[-] closed code=${code} reason=${reason}`);
    });

    ws.on('error', (err) => {
      console.error(`[!] error: ${err.message}`);
    });
  });
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: tools/ws_test_server.py
#!/usr/bin/env python3
"""
Minimal WebSocket test server for mod_audio_stream.

- Logs all incoming text and binary messages.
- Echoes text back with a small JSON envelope.
- Optionally sends a streamAudio JSON with a provided raw PCM file (16-bit little-endian).

Usage:
  python3 tools/ws_test_server.py --port 8765
  python3 tools/ws_test_server.py --port 8765 --send-raw sample.raw --rate 8000 --channels 1

Requires:
  pip install websockets
"""
import argparse
import asyncio
import base64
import json
import pathlib
import websockets


async def handler(websocket, path, opts):
    peer = websocket.remote_address
    print(f"[+] connection from {peer}")

    # Optionally send one streamAudio message on connect
    if opts.send_raw:
        try:
            raw = pathlib.Path(opts.send_raw).read_bytes()
            payload = {
                "type": "streamAudio",
                "data": {
                    "audioDataType": "raw",
                    "sampleRate": opts.rate,
                    "audioData": base64.b64encode(raw).decode("ascii"),
                },
            }
            await websocket.send(json.dumps(payload))
            print(f"[>] sent streamAudio: bytes={len(raw)} rate={opts.rate}")
        except Exception as exc:
            print(f"[!] failed to send raw audio: {exc}")

    try:
        async for message in websocket:
            if isinstance(message, str):
                print(f"[<] text: {message[:200]}")
                reply = {"echo": message}
                await websocket.send(json.dumps(reply))
                print(f"[>] echo reply")
            else:
                print(f"[<] binary: {len(message)} bytes")
                # no echo for binary to avoid flooding
    except websockets.ConnectionClosed as e:
        print(f"[-] connection closed: code={e.code} reason={e.reason}")
    except Exception as exc:
        print(f"[!] handler exception: {exc}")


def main():
    parser = argparse.ArgumentParser(description="Test WebSocket server for mod_audio_stream")
    parser.add_argument("--port", type=int, default=8765, help="Port to listen on")
    parser.add_argument("--host", default="0.0.0.0", help="Host to bind")
    parser.add_argument("--send-raw", help="Path to raw PCM file (16-bit little-endian) to send once on connect")
    parser.add_argument("--rate", type=int, default=8000, help="Sample rate for the raw PCM (used in JSON metadata)")
    parser.add_argument("--channels", type=int, default=1, help="Channels for the raw PCM (unused by client but informational)")
    args = parser.parse_args()

    async def ws_handler(ws, path):
        await handler(ws, path, args)

    start_server = websockets.serve(ws_handler, args.host, args.port)
    print(f"[*] listening on ws://{args.host}:{args.port}")
    asyncio.get_event_loop().run_until_complete(start_server)
    asyncio.get_event_loop().run_forever()


if __name__ == "__main__":
    main()
